### ISSUE TITLE
Quasi 2D-3D coupling fails for 2D SU2 grid and 3D CalculiX grid

### ISSUE THREAD

**Issue by SangeetSunderroy** [precice/calculix-adapter#61]
Hello,
System:
- Ubuntu 20.04
- Precice 2.2.0 installed using the .deb file for Ubuntu 20.04. 
- CalculiX 2.16
- SU2 6.0.0
I currently have a 2D SU2 mesh and I have a 3D CalculiX mesh which has only 1 element in the z-direction. So, I have set the following in my precice-config file.

`<solver-interface dimensions="2">`

 But when I run CalculiX using the command

`ccx_preCICE -i ramp -precice-participant Solid`

I get the following error
```
Using quasi 2D-3D coupling
Set ID Found 
ccx_preCICE: adapter/PreciceInterface.c:559: PreciceInterface_ConfigureNodesMesh: Assertion `count == interface->num2DNodes && "Filtering of 2D nodes not done properly. Please make sure the out-of-plane axis is Z-axis"' failed.
```
But my CalculiX grid's out of plane axis is the z-axis. Do you have any suggestions on what I could be doing incorrectly? I have attached my case folder.
[precice_25_degree_compression_ramp.zip](https://github.com/precice/calculix-adapter/files/6352576/precice_25_degree_compression_ramp.zip)

**Comment by IshaanDesai** [comment]
Hi @SangeetSunderroy ,
I checked your case and realize that you have used second order elements while meshing CalculiX, this essentially leads to two two additional layers in the Z-direction. One layer are the corner nodes of the element and the other layer is the intermediate higher order node on the edge of the element in the Z-direction. The current quasi 2D-3D mechanism is not capable of filtering higher order nodes and hence fails to work for your case. 
One solution would be to mesh the geometry with first order elements. If you want to have higher order elements and still use the quasi 2D-3D mechanism then have a look at [this issue](https://calculix.discourse.group/t/two-types-of-elements-in-a-cgx-mesh/483/4). It does not offer a full solution but a way to try things out. 
If you successfully run the quasi 2D-3D case using higher order elements then we would be very interested in knowing the strategy you pursued.

**Comment by AndresPedemonteFIUBA** [comment]
I'm running into the same problem even though my mesh uses linear elements (C3D8). Mesh is appropriate for a quasi 2D-3D case: geometry lies in the XY plane with one element in the Z-direction (goes from Z=0 to Z=1).

The following is set in preCICE config file:
`<solver-interface dimensions="2">`
and I get the same error the original poster received:
`---[precice]  Revision info: v2.3.0
---[precice]  Configuration: Release (Debug and Trace log unavailable)
---[precice]  Configuring preCICE with configuration "../precice-config.xml"
---[precice]  I am participant "Solid"
Using quasi 2D-3D coupling
Set ID Found 
ccx_preCICE: adapter/PreciceInterface.c:559: PreciceInterface_ConfigureNodesMesh: Assertion `count == interface->num2DNodes && "Filtering of 2D nodes not done properly. Please make sure the out-of-plane axis is Z-axis"' failed.`

Of particular note is that, for this case, I have split the FSI interface into three (top, bottom and side of a very thin flap). However, this does not seem to be the root of the issue as I have tested the same case modified to have only one interface and still receive the same error.

There is no node in the mesh that has a Z-coordinate different than 0 and 1. I have also double-checked the interface node sets, so the issue does not stem from that. All elements are first-order (linear) elements.

System info:
Ubuntu 20.04
preCICE 2.3.0
CalculiX 2.16 + latest adapter

**Comment by boris-martin** [comment]
Hi @AndresPedemonteFIUBA, 
I plan to investigate this issue on the code. Would you mind sharing your case to help me reproduce the issue ?

**Comment by AndresPedemonteFIUBA** [comment]
Hey @boris-martin,

Yes, I'm attaching it here. It's a very simple mesh, with 70x1x1 (XYZ) elements. That is, every node is part of the interface. 

You'll find 2 directories in the .zip file: `one-interface,` `three-interfaces`. Each one has an appropriate `precice-config.xml`. Both have the same mesh.

In the first one (`one-interface`), there is only one interface made up from the node set containing all nodes. No error here.

In the second one, the fluid/solid interface is split into three. There, nodes 1, 72, 143 and 214 make up the flap's tip (side surface). Nodes 1 and 72 are part of two interfaces,`surfaceBottom` and `surfaceSide`, while nodes 143 and 214 are part of both `surfaceTop` and `surfaceSide`. The error shows up when running this `three-interfaces` case.

I've noticed something interesting though, which I added as a comment in the `mesh.msh` file of `three-interfaces`. If the node set `NsurfaceSide` is defined with
```
**NSET, NSET=NsurfaceSide
  1,   72,   143,  214
```
instead of
```
*NSET, NSET=NsurfaceSide, generate
   1,   214,  71
```
The error message does not show up (at least not when starting the simulation; I have not tested coupling it with another participant after this change). However, both ways of defining the node set are equivalent (or should be in theory).

Here it is:
[2d3d-foil.zip](https://github.com/precice/calculix-adapter/files/7926552/2d3d-foil.zip)

Let me know if you need anything else from me.

Andrés

**Comment by AndresPedemonteFIUBA** [comment]
If it's any help to know, I tried coupling it with another participant after doing that "trick" that allowed me to get past the failed assertion but it still fails during the first coupling iteration with a segmentation fault. I can't say with certainty that the split interface is the cause of that segfault, since I get the same if I run the case with only one interface (I reduced it to the most basic set-up: one interface, NN mapping, etc.). That same set-up works with ccx 2.15 +adapter though (of course, with `<solver-interface dimensions="3">` for 2.15) .

Andrés

**Comment by boris-martin** [comment]
Hi @AndresPedemonteFIUBA ,

I've been working on a big overhaul of 2D-3D coupling (https://github.com/precice/calculix-adapter/pull/92) and it seems to fix your previous issues (issue #86 is still there). I could run both your cases (from the files sent on Jan. 24) with that version. (However, *both* failed with the current released version, which surprised me)
Would you like to give it a try? Can I close this issue once its gets merged?

> - Did not use the generate parameter for any node set definition. All node sets are defined by spelling out each and every node that makes them up.

I have zero knowledge of this "generate" function. I always used explicit spelling.

> - I renamed the node sets from NsurfaceTop to NTop, NsurfaceBottom to NBottom and NsurfaceSide to NSide.

Do you think starting with a capital is important ? AFAIK it shouldn't.

> - Each interface node set has its own .nam file for its definition. That means each *NSET, NSET=... is at the top of its .nam file. I was inspired to try this after reading a message by @KyleDavisSA on the preCICE Gitter chatroom from May 30 2020 where he stated: []

Actually all files are included in the .inp file, the .nam file is a made up convention. But the NSET must match must indeed match everywhere.

**Comment by boris-martin** [comment]
> > Actually all files are included in the .inp file, the .nam file is a made up convention. But the NSET must match must indeed match everywhere.
> 
> Yes, but the key here was that each patch (`NSET` here) of the fluid/solid interface (remember I split it into 3) must have its definition in its own `.nam` file. If I defined all of them in a single `.nam` file, I'd get the filtering error.

That's weird, I'd expect the result to be independent as long as the includes are properly done.
